src/json_schema/keywords/pattern.rs

```diff
@@ -12,7 +12,8 @@ impl super::Keyword for Pattern {
 
         if pattern.is_string() {
             let pattern_val = pattern.as_str().unwrap();
-            match regex::Regex::new(pattern_val) {
+            let pattern_val_fixed = str::replace(pattern_val, r"\/", "/");
```

Comment:
Hmm, this smells like trouble somehow.

[JSON Schema](https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.6.4) refers to [ECMA262](https://www.ecma-international.org/publications/files/ECMA-ST/Ecma-262.pdf) for the definition of a regular expression (incorrectly linking to a section that does not exist — it is actually in §11.8.5), which in turn mandates that

- `\`, `/`, and `[` (as well as line endings) are not legal terminals in the expression
- and that everything that is not a line ending may be escaped

Rust’s `regex` OTOH only permits [explicitly specified](https://docs.rs/regex-syntax/0.6.18/src/regex_syntax/ast/parse.rs.html#1534) escape sequences to be started by the `\` character. So the full solution would entail actually parsing the regex according to the ECMA grammar and then emitting a valid expression according to the Rust grammar — which definitely sounds like overkill. But with the patch above I worry that `\\/` (which is not ECMA compliant, but may otherwise work) would be incorrectly changed into `\/`; in this case that’s okay because the result doesn’t pass `Regex::new` … we might be fine here, but it feels dirty.

Reply:
> I worry that `\\/` (which is not ECMA compliant, but may otherwise work) would be incorrectly changed into `\/`; in this case that’s okay because the result doesn’t pass `Regex::new` … we might be fine here, but it feels dirty.

What do you mean by this? `\\/` is a valid JavaScript RegExp:
```
new RegExp('\\/').test('/')
> true
```

which is parsed (by serde) as `\/` and needs to be fed into `Regex::new` as `/`.

Or do you mean the already escaped version resulting from `new RegExp('\\\\/')`? That seems to indeed break this logic.

I agree that the fix looks dirty but I think it's the right place to put it: Where the JSON regex is transformed into a Rust regex (with different escaping rules).

Maybe we could get away with switching to https://github.com/ridiculousfish/regress which after a quick test seems to do the right thing.

Reply:
What feels dirty to me is that this fixes exactly one problematic case, namely the unescapable one where JSON Schema demands something that `regex` forbids. The other problematic cases are all escaped unicode codepoints that are _not_ escape sequences in JSON Schema, which need to have their backslash stripped to be treated correctly by our Rust code.

Reply:
OTOH the JSON Schema spec itself is vague on which escape sequences should be supported, which is … unfortunate (i.e. when you write a pattern with escape sequences, you cannot be sure whether those will be treated as the unicode codepoint or specially, since that is implementation dependent).

Reply:
> Maybe we could get away with switching to https://github.com/ridiculousfish/regress which after a quick test seems to do the right thing.

[I gave this a try.](https://github.com/Actyx/valico/pull/3)

Reply:
Hmm, that’s quite invasive. As JSON (like everything JS) is inherently “unsupportable” because it is ill-defined, we may also just merge this PR and then document that only the Rust `regex` escape sequences plus `\/` are valid.